docs/rationale.md

```diff
@@ -0,0 +1,144 @@
+
+## Large tests
+
+Test262 tests are typically very focused. The vast majority exercise just one
+algorithm step/grammar production, and some are even more granular that that!
+
+Some test contributors are uncomfortable splitting their work across files like
+this. It's certainly unlike the practices that are common in modern application
+development. In those settings, many tests are often grouped into the same file
+and separated by function boundaries.
+
+Test262 doesn't use the same approach as a typical application test suite in
+order to limit complexity. The guidance of "one test per file" means that
+consuming Test262 is relatively easy; there is no "test runner" API for
+consumers to implement, and interpreting results is likewise straightforward.
+It also lowers the barrier to entry for new contributors since there is no API
+to learn.
+
+## Syntax tests
+
+When testing a syntactic feature of the language, it can be tempting to write
+tests which verify that some bit of source text does *not* produce a syntax
+error. Contributors should try to push beyond verifying only the lack of a
+syntax error because almost all such tests also have observable semantics. It's
+often better for a test to assert that the expected semantics are followed,
+even when they may already be covered elsewhere.
+
+However, this is not always desirable because verifying semantics invariably
+requires inserting still more code, and that additional code may degrade the
+tests' precision for verifying syntax.
+
+When considering this tension, be aware that TC39 maintains [a separate project
+called test262-parser-tests](https://github.com/tc39/test262-parser-tests).
+This project was partially motivated by a desire to offer test material that
+isn't (and perhaps cannot be) related to any specific grammar production. The
+availability of that project may inform decisions about if, where, and how to
+include tests for syntax in Test262.
+
+## Avoiding abstraction
+
+Contributors will occasionally suggest introducing new abstractions to reduce
+duplication in tests. The maintainers set a relatively high bar for such
+enhancements, both due to their many drawbacks and due to the aspects of
+standards testing which limit their benefit.
+
+The drawbacks to abstraction include:
+
+- it degrades the tests by introducing unrelated semantics
+- it discourages contributors by requiring them to learn more
+- it frustrates implementers by making it harder to understand what's being
+  tested and what has failed
+
+One of abstraction's common motivations is its tendency to reduce maintenance
+costs by limiting duplication. TC39 has a very high standard for compatibility
+between revisions of ECMA262. This gives us a certain assurance in Test262 that
+maintainers of other test suites do not enjoy: Test262's tests are very rarely
+invalidated. The project takes advantage of this by using a more declarative,
+readable, and verbose style.
```

Comment:
I'm OK with setting a high bar for abstractions for the reasons you give here, and it is also current practice. But I don't agree philosophically with some of the drawbacks (like abstractions being less readable), or I feel that the corresponding advantages aren't really mentioned. So I'm not sure I'd feel comfortable about enshrining that as policy. I do agree for sure that there is a line to be drawn, so maybe we could phrase this differently?

E.g., I'd be in favour of removing these two paragraphs, and having the last paragraph say something like "Abstraction has other motivations, so there will always be room for it to some extent. We expect in the future to have more guidance about when it is appropriate and when it isn't. When the benefits of a specific proposal ..."

For example, I feel that the part about test262 tests rarely being invalidated holds very well for testing things that are already part of the standard, but not for developing tests for pre-Stage-3 proposals, which was my main exposure to test262 for a long time. So that may help illustrate why I'm approaching this from a different perspective.

Reply:
This reflects our current thinking. It's certainly not set in stone, though, and I'm particularly interested in your experience as a proposal champion. At the same time, I don't think this pull request is well-suited to making changes to the rationale. Starting a dedicated discussion based on a finalized document (perhaps using the nascent RFC process) would promote visibility and make it easier for others to recognize if/how maintenance practices changed over time.

Given that the proposed text doesn't give any concrete instruction on the topic, how do you feel about merging it as-is and following up with a discussion about improvements?